### PR TITLE
UI Redesign

### PR DESCRIPTION
**Pull request by Grarak**
- Change primary colors to white/black for light/dark respectively
- Accent color stays red
- Replace refresh icon with pull to refresh
- Replace toolbar with custom search bar
- Change grid design layout
- Basic implementation of ViewModel
- Usage of kotlin coroutines
- Make OSC visible if controller id 0 is enabled

### CODE REVIEW

**Comment by Dyldaboss** [comment]
Where do I get the NRO file

**Comment by FAROK000** [comment]
Grarak
Ammn the control buttons are little bit annoying when moving a, b, x, y alone in the screen  the same for d-pad
It must be look like  ppsspp (all move together).
Just an opinion 😊

**Review by PixelyIon — APPROVED** [review]
After an extensive review of the code, I've determined yet again that you are better at Kotlin than me and it even being reviewed by me is pointless.

**Review by IvarWithoutBones — APPROVED** [review]
The redesign looks beautiful, great job!

**Comment by ezequielbsb25** [comment]
> The redesign looks beautiful, great job!
Indeed, but after the redesign update the Sonic xci was didn't find/recognized by the emulator

**Comment by PixelyIon** [comment]
XCIs do work fine on my end so it could be a local issue, head on over to our [Discord](https://discord.gg/XnbXNQM) so we can resolve the issue.

**Comment by IvarWithoutBones** [comment]
This has been resolved on Discord, @ezequielbsb25 had a dump with a wrong file extension causing the emulator to not detect it.

File: app/src/main/java/emu/skyline/KeyReader.kt [hunk 1]
```diff
@@ -26,60 +26,53 @@ object KeyReader {
...
-                        if (value.length != 32 && !isHexString(value))
-                            return@useLines false
-                    }
-                    KeyType.Prod -> {
-                        if (!key.contains("_"))
-                            return@useLines false
-                        if (!isHexString(value))
-                            return@useLines false
+        tmpOutputFile.bufferedWriter().use { writer ->
+            val valid = inputStream!!.bufferedReader().useLines {
```
Comment (sspacelynx) [inline app/src/main/java/emu/skyline/KeyReader.kt:38]: Is the not-null assertion operator avoidable here? I'd avoid it in all cases when possible, maybe by adding some error handling and displaying a Toast/Snackbar to the user if necessary.

File: app/src/main/java/emu/skyline/RomProvider.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                addEntries(fileFormats, file, entries)
+            } else {
+                fileFormats[file.name?.substringAfterLast(".")?.toLowerCase()]?.let { romFormat->
+                    entries.getOrPut(romFormat, { arrayListOf() }).add(RomFile(context, romFormat, file.uri).appEntry)
+                }
+            }
+        }
+    }
+
+    fun loadRoms(searchLocation : Uri) = DocumentFile.fromTreeUri(context, searchLocation)!!.let { documentFile ->
```
Comment (sspacelynx) [inline app/src/main/java/emu/skyline/RomProvider.kt:33]: Same thing about the not-null assertion operator here.

File: build.gradle [hunk 3]
```diff
@@ -1,27 +1,33 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
-    ext.kotlin_version = '1.4.30'
+    ext {
+        kotlin_version = '1.4.30'
```
Comment (sspacelynx) [inline build.gradle:5]: Nitpick here, kotlin can be bumped to 1.5 but I assume you were probably going to do that anyway.

**Review by sspacelynx — COMMENTED** [review]
Code looks awesome, your Kotlin is definitly better than mine too.
I'd avoid the not-null assertion operator at all costs, there's a few that I did not report individually in the controllers' code. They might trigger only in rare situations, but I still think they should be handled appropriately.

File: app/src/main/java/emu/skyline/KeyReader.kt (hunk 1 above)
Comment (Grarak) [reply app/src/main/java/emu/skyline/KeyReader.kt:38]: `inputStream` can only be `null` when the content provider has crashed. This shouldn't happen, unless there is something wrong with the OS

File: app/src/main/java/emu/skyline/RomProvider.kt (hunk 2 above)
Comment (Grarak) [reply app/src/main/java/emu/skyline/RomProvider.kt:33]: `fromTreeUri` can only return `null` when the target API is lower than 21

File: build.gradle (hunk 3 above)
Comment (Grarak) [reply build.gradle:5]: I'll do that later, `1.5` wasn't released when I've started with this PR

**Review by PixelyIon — APPROVED** [review]
You've really improved the UI so dramatically that it's up there with the best UIs I've ever seen, genuinely amazing job! :tada:

**Pull request merged**